ksql-engine/src/main/java/io/confluent/ksql/topic/DefaultTopicInjector.java

```diff
@@ -0,0 +1,202 @@
+import io.confluent.ksql.parser.tree.IntegerLiteral;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.StringLiteral;
+import io.confluent.ksql.services.KafkaTopicClient;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlConstants;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlStatementException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.clients.admin.TopicDescription;
+
+public class DefaultTopicInjector implements TopicInjector {
+
+  private final KafkaTopicClient topicClient;
+  private final MetaStore metaStore;
+
+  public DefaultTopicInjector(
+      final KsqlExecutionContext executionContext
+  ) {
+    this(executionContext.getServiceContext().getTopicClient(), executionContext.getMetaStore());
+  }
+
+  DefaultTopicInjector(
+      final KafkaTopicClient topicClient,
+      final MetaStore metaStore) {
+    this.topicClient = Objects.requireNonNull(topicClient, "topicClient");
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> PreparedStatement<T> forStatement(
+      final PreparedStatement<T> statement,
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> propertyOverrides) {
+    if (!(statement.getStatement() instanceof CreateAsSelect)) {
+      return statement;
+    }
+
+    final PreparedStatement<? extends CreateAsSelect> cas =
+        (PreparedStatement<? extends CreateAsSelect>) statement;
+
+    final String topic = topicName(cas.getStatement(), ksqlConfig);
+    final TopicDescription description = describeSource(topicClient, ksqlConfig, cas);
+    final int partitions = numPartitions(cas, ksqlConfig, propertyOverrides, description);
+    final short replicas = numReplicas(cas, ksqlConfig, propertyOverrides, description);
+
+    final Map<String, Expression> properties = new HashMap<>(cas.getStatement().getProperties());
+    properties.put(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY, new StringLiteral(topic));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_REPLICAS, new IntegerLiteral(replicas));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_PARTITIONS, new IntegerLiteral(partitions));
+
+    final CreateAsSelect withTopic = cas.getStatement().copyWith(properties);
+    final String withTopicText = SqlFormatter.formatSql(withTopic) + ";";
+
+    return (PreparedStatement<T>) PreparedStatement.of(withTopicText, withTopic);
+  }
+
+  private TopicDescription describeSource(
+      final KafkaTopicClient topicClient,
+      final KsqlConfig ksqlConfig,
+      final PreparedStatement<? extends CreateAsSelect> cas
+  ) {
+    final Analysis analysis = new QueryAnalyzer(
```

Comment:
I still strongly feel this is a misuse of `QueryAnalyzer`.  `QueryAnalyzer` will be enhanced as we move forward to be more complex, more advanced, and hence take longer.  Using it here to extract a single property of a query is total overkill, coupling we don't need, and ultimately costly. 

The query has already been analysed higher up the stack by the time this method is called. We should not be analysing it again. 

I see two approaches to remove the need for this:

1. Capture this information higher up the stack when the query is analysed and pass it down.
2. Write a very simple visitor to extract the info, e.g.

```java
private TopicDescription describeSource(
      final KafkaTopicClient topicClient,
      final KsqlConfig ksqlConfig,
      final PreparedStatement<? extends CreateAsSelect> cas
  ) {
    final SourceTopicExtractor visitor = new SourceTopicExtractor();
    visitor.process(cas.getStatement().getQuery(), null);

    if (visitor.primaryKafkaTopicName == null) {
      throw new IllegalStateException("blag");
    }
    return topicClient.describeTopic(visitor.primaryKafkaTopicName);
  }

  private final class SourceTopicExtractor extends DefaultTraversalVisitor<Node, Void> {

    private String primaryKafkaTopicName = null;

    @Override
    protected Node visitJoin(final Join node, final Void context) {
      process(node.getLeft(), context);
      return null;
    }

    @Override
    protected Node visitAliasedRelation(final AliasedRelation node, final Void context) {
      final String structuredDataSourceName = ((Table) node.getRelation()).getName().getSuffix();
      final StructuredDataSource<?> source = metaStore.getSource(structuredDataSourceName);
      if (source == null) {
        throw new KsqlException(structuredDataSourceName + " does not exist.");
      }

      primaryKafkaTopicName = source.getKsqlTopic().getKafkaTopicName();
      return node;
    }
  }
```

With regards to future enhancements, changes to the `QueryAnalyser` etc... If we ensure this injector is called as part of the `QueryTranslationTest` then it should be future proofed.

Reply:
Ok. I will go with approach 2.